### PR TITLE
linkable gene searches

### PR DESCRIPTION
**Pull request by devonbush**
Add url-routable gene searches, and moves the gene search capability over to React/API. 
One question: right now I've put react-owned routes under `/single_cell/app/`.   So e.g.  `http://localhost:3000/single_cell/app/genes` routes to the gene search tab.   Is this the right place?  Should we just route it to /app to make urls shorter?  Or is the `single_cell` prefix valuable for other reasons?

image

This does need backend tests, I will add those tomorrow, but wanted to get at least the frontend out for Eric and Eno to start their reviews.

### CODE REVIEW

**Comment by bistline** [comment]
The `/single_cell` prefix use to be required when we were on the Broad load balancer, but now it is less critical.  Changing it however would be a major refactoring, so let's skip that for now.  I think `app/genes` makes sense for now.

File: db/migrate/20200413211749_add_feature_flag_defaults.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+
+
+
+
+class AddFeatureFlagDefaults < Mongoid::Migration
+  # mirror of FeatureFlag.rb, so this migration won't error if that class is renamed/altered
+  class FeatureFlagMigrator
```
Comment (bistline) [inline db/migrate/20200413211749_add_feature_flag_defaults.rb:7]: Very cool!

File: app/controllers/api/v1/expression_data_controller.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      before_action :set_study
+      before_action :check_study_view_permission
+
+
+      # Returns the specified expression data for the gene within the given study, optimized for rendering
+      # by the SCP UI.
+      # We agreed that there would be no swagger docs for this endpoint, as it is not intended
+      # to be used other than by the SCP UI, and may change dramatically
+      def show
+        data_type = params[:id]
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/controllers/api/v1/expression_data_controller.rb:18]: agree, didn't realize rails let you customize this--will change

File: app/controllers/api/v1/expression_data_controller.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        if (data_type == 'violin')
+          render_violin
+        elsif (data_type == 'annotations')
+          render_annotation_values
+        else
+          render json: {error: "Unknown expression data type: #{data_type}"}, status: 404
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/controllers/api/v1/expression_data_controller.rb:24]: yup

File: app/lib/study_search_service.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    gene_matches = Gene.where(:study_id.in => study_ids)
+                       .any_of({:name.in => genes},
+                               {:searchable_name.in => genes.map(&:downcase)},
+                               {:gene_id.in => genes})
+    gene_match_list = gene_matches.pluck(:id, :searchable_name, :study_id)
+    genes_by_study = {}
+    study_ids = []
+    gene_match_list.map do |match|
+      genes_by_study[match[2]] = genes_by_study[match[2]].present? ? genes_by_study[match[2]].push(match[1]) : [match[1]]
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/lib/study_search_service.rb:29]: ooh, that is a nice formulation, concur

File: app/javascript/components/search/genes/StudyGeneDotPlot.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  const [isLoading, setIsLoading] = useState(false)
+  /** fetch the expression data from the server */
+  async function loadData() {
+    setIsLoading(true)
+    const annotations = await fetchAnnotationValues(study.accession)
+    if (study.gene_matches.length > 1) {
+      const geneParam = genes.join('+')
+      window.renderMorpheusDotPlot(
+        `/single_cell/study/${study.accession}/${studyNameAsUrlParam(study.name)}/expression_query?search[genes]=${geneParam}&row_centered=&row_centered=&cluster=&request_user_token=${userState.accessToken}`,
+        `/single_cell/study/${study.accession}/${studyNameAsUrlParam(study.name)}/annotation_query?cluster=&annotation=CLUSTER--group--study&request_user_token=${userState.accessToken}`,
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/javascript/components/search/genes/StudyGeneDotPlot.js:22]: good catch, yeah, this is where the annotation being called 'CLUSTER' threw me off

**Comment by eweitz** [comment]
Also, please add GitHub `security-risk` label.

File: app/controllers/api/v1/expression_data_controller.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      end
+
+      def render_violin
+        cluster = ApiParamUtils.get_cluster_group(params, @study)
+        selected_annotation = ApiParamUtils.get_selected_annotation(params, @study, cluster)
+        subsample = params[:subsample].blank? ? nil : params[:subsample].to_i
+        gene = @study.genes.by_name_or_id(params[:gene], @study.expression_matrix_files.map(&:id))
+
+        render_data = ExpressionRenderingService.get_global_expression_render_data(
+          @study, subsample, gene, cluster, selected_annotation, params[:boxpoints], current_user
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/controllers/api/v1/expression_data_controller.rb:36]: done

File: app/controllers/api/v1/search_controller.rb [hunk 7]
```diff
@@ -41,6 +42,13 @@ class SearchController < ApiBaseController
             key :required, false
             key :type, :string
           end
+          parameter do
+            key :name, :genes
+            key :in, :query
+            key :description, 'User-supplied query string'
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/controllers/api/v1/search_controller.rb:48]: done

File: app/javascript/components/covid19/Covid19PageContent.js [hunk 8]
```diff
@@ -10,6 +12,14 @@ import ErrorBoundary from 'lib/ErrorBoundary'
...
  */
 export default function Covid19PageContent() {
+  return (
+    <Router>
+      <UnroutedPageContent default/>
+    </Router>
+  )
+}
+
+function UnroutedPageContent() {
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/javascript/components/covid19/Covid19PageContent.js:22]: done

File: app/javascript/lib/scp-api-metrics.js [hunk 9]
```diff
@@ -89,7 +92,7 @@ export function logSearch(type, searchParams) {
 
   const numTerms = getNumberOfTerms(terms)
   const [numFacets, numFilters] = getNumFacetsAndFilters(facets)
-  const facetList = Object.keys(facets)
+  const facetList = facets ? Object.keys(facets) : {}
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/javascript/lib/scp-api-metrics.js:95]: great catch, thanks

File: app/javascript/components/search/genes/StudyGeneDotPlot.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  const [isLoaded, setIsLoaded] = useState(false)
+  const [isLoading, setIsLoading] = useState(false)
+  /** fetch the expression data from the server */
+  async function loadData() {
+    setIsLoading(true)
+    const annotations = await fetchAnnotationValues(study.accession)
+    if (study.gene_matches.length > 1) {
+      const geneParam = genes.join('+')
+      window.renderMorpheusDotPlot(
+        `/single_cell/study/${study.accession}/${studyNameAsUrlParam(study.name)}/expression_query?search[genes]=${geneParam}&row_centered=&row_centered=&cluster=&request_user_token=${userState.accessToken}`,
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/javascript/components/search/genes/StudyGeneDotPlot.js:21]: yup, I'll add this as a topic for sprint planning

File: app/javascript/components/search/genes/GeneSearchView.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+  const featureFlagState = useContext(FeatureFlagContext)
+  const geneSearchState = useContext(GeneSearchContext)
+  const studySearchState = useContext(StudySearchContext)
+  const [genes, setGenes] = useState(_clone(geneSearchState.params.genes))
+  const [showStudyControls, setShowStudyControls] = useState(hasSearchParams(studySearchState.params))
+
+  /** handles a user submitting a gene search */
+  function handleSubmit(event) {
+    event.preventDefault()
+    geneSearchState.updateSearch({ genes }, studySearchState, featureFlagState.gene_study_filter)
```
[the remark this answers is not in the document]
Comment (devonbush) [reply app/javascript/components/search/genes/GeneSearchView.js:28]: yeah, it's definitely a topic worth some thought.

**Pull request merged**